dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java

```diff
@@ -3474,6 +3485,122 @@ protected void checkLogsForErrorsAndWarnings(List<String> failuresRegExps, Strin
+
+                Set<String> expectedFeatures = new HashSet<>(featureReplacementAction.getAddFeatures()); //the expected features, if present
+                expectedFeatures.addAll(featureReplacementAction.getAlwaysAddFeatures());
+                Set<String> installedFeatures = getInstalledFeatures(); //the features actually installed at runtime
+
+                //compare each installed feature to the expected ones
+                for (String installedFeature : installedFeatures) {
+                    //ignore versionless features
+                    int dash = installedFeature.indexOf("-");
+                    if (dash > -1) {
+                        String versionlessInstalledFeature = getVersionlessFeatureName(installedFeature);
+                        for (String replacementFeature : expectedFeatures) {
+                            String versionlessReplacementFeature = getVersionlessFeatureName(replacementFeature);
+                            if (versionlessReplacementFeature.equalsIgnoreCase(versionlessInstalledFeature)) {
+                                //if the base feature name matches but the version does not, throw an exception.
+                                if (!replacementFeature.equalsIgnoreCase(installedFeature)) {
+                                    String message = "Runtime feature was not of the expected version for repeat action (" + serverName + ", " + action.getID() + "). "
+                                                     + "Expected: " + replacementFeature + ", Actual: " + installedFeature + ". "
+                                                     + "This is caused by a feature not being explicitly set in the FAT's server.xml such that FeatureReplacementAction does not replace it properly.";
+
+                                    //check for exempt servers should eventually be removed
+                                    //if exempt then output info message, otherwise throw exception
+                                    if (EXEMPT_SERVERS_SET.contains(serverName)) {
+                                        Log.info(c, method, message);
+                                    } else {
+
+                                        throw new Exception(message);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        //re-instate this message once the exempt servers are removed
+        //Log.info(c, method, "No invalid replacement features found.");
+    }
+
+    /**
+     * Extract only the versionless base name of a feature, everything up to the dash.
+     * If the feature is already versionless (no dash) return as is.
+     *
+     * @param  feature The short feature name
+     * @return         The versionless feature name
+     */
+    public static final String getVersionlessFeatureName(String feature) {
```

Comment:
Can't decide whether this use of "versionless" is a good idea, or confusing given the existence of "versionless features".

Tempted to rename to `removeFeatureVersion(String featureName)`